controllers/datasciencecluster/datasciencecluster_controller.go

```diff
@@ -60,486 +60,325 @@ type DataScienceClusterReconciler struct {
-		//	}
-		// }
-		if controllerutil.ContainsFinalizer(instance, finalizerName) {
-			controllerutil.RemoveFinalizer(instance, finalizerName)
-			if err := r.Update(ctx, instance); err != nil {
-				return ctrl.Result{}, err
-			}
-		}
-		if upgrade.HasDeleteConfigMap(ctx, r.Client) {
-			// if delete configmap exists, requeue the request to handle operator uninstall
-			return reconcile.Result{Requeue: true}, nil
-		}
+	return ctrl.Result{}, nil
+}
 
-		return ctrl.Result{}, nil
+func (r *DataScienceClusterReconciler) validate(ctx context.Context, _ *dscv1.DataScienceCluster) error {
+	// This case should not happen, since there is a webhook that blocks the creation
+	// of more than one instance of the DataScienceCluster, however one can create a
+	// DataScienceCluster instance while the operator is stopped, hence this extra check
+
+	dscInstances := &dscv1.DataScienceClusterList{}
+	if err := r.Client.List(ctx, dscInstances); err != nil {
+		return fmt.Errorf("failed to retrieve DataScienceCluster resource: %w", err)
 	}
 
-	// Start reconciling
-	if instance.Status.Conditions == nil {
-		reason := status.ReconcileInit
-		message := "Initializing DataScienceCluster resource"
-		instance, err = status.UpdateWithRetry(ctx, r.Client, instance, func(saved *dscv1.DataScienceCluster) {
-			status.SetProgressingCondition(&saved.Status.Conditions, reason, message)
-			saved.Status.Phase = status.PhaseProgressing
-			saved.Status.Release = currentOperatorRelease
-		})
-		if err != nil {
-			_ = r.reportError(ctx, err, instance, fmt.Sprintf("failed to add conditions to status of DataScienceCluster resource name %s", req.Name))
+	if len(dscInstances.Items) != 1 {
+		return fmt.Errorf("failed to get a valid DataScienceCluster instance, expected to find 1 instance, found %d", len(dscInstances.Items))
+	}
 
-			return ctrl.Result{}, err
-		}
+	dsciInstances := &dsciv1.DSCInitializationList{}
+	err := r.Client.List(ctx, dsciInstances)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve DSCInitialization resource: %w", err)
 	}
 
+	if len(dsciInstances.Items) != 1 {
+		return fmt.Errorf("failed to get a valid DSCInitialization instance, expected to find 1 instance, found %d", len(dscInstances.Items))
+	}
+
+	return nil
+}
+
+func (r *DataScienceClusterReconciler) reconcileComponents(ctx context.Context, instance *dscv1.DataScienceCluster) error {
+	log := logf.FromContext(ctx).WithName("DataScienceCluster")
+
+	notReadyComponents := make([]string, 0)
+
 	// all DSC defined components
 	componentErrors := cr.ForEach(func(component cr.ComponentHandler) error {
-		var err error
-		instance, err = r.ReconcileComponent(ctx, instance, component)
-		return err
+		ci, err := r.reconcileComponent(ctx, instance, component)
+		if err != nil {
+			return err
+		}
+
+		readyCondition := meta.FindStatusCondition(ci.GetStatus().Conditions, status.ConditionTypeReady)
+		if readyCondition != nil && readyCondition.Status != metav1.ConditionTrue {
```

Comment:
Is it ready if there is not ready condition?

Reply:
good spot, let me fix it

Reply:
Pushed a commit, now the logic is that:
- if the component is not `managed`, skip checking its readiness
- if the component is `managed`, then it is ready if the ready condition is present with status `True`
- the message DSC's `Ready` condition message has been made generic as multiple factors could contribute to set it to ready (i.e. a cluster without managed components is still `Ready`)